keps/sig-cluster-lifecycle/0035-20190128-addons-via-operators.md

```diff
@@ -0,0 +1,222 @@
+---
+kep-number: 35
+title: Addons via Operators
+authors:
+  - "@justinsb"
+owning-sig: sig-cluster-lifecycle
+reviewers:
+  - @luxas
+  - @roberthbailey
+  - @timothysc
+approvers:
+  - @timothysc
+editor: TBD
+creation-date: 2019-01-28
+last-updated: 2019-03-11
+status: provisional
+---
+
+# Addons via Operators
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Infrastructure Needed](#infrastructure-needed)
+
+
+## Summary
+
+We propose to use operators for managing cluster addons.  Each addon will have
+its own CRD, and users will be able to perform limited tailoring of the addon
+(install/don’t install, choose version, primary feature selection) by modifying
+the CR.  The operator encodes any special logic (e.g. dependencies) needed to
+install the addon.
+
+By creating a CRD per addon, we are able make use of the kubernetes API
+machinery for those per-addon options.
+
+We will create tooling to make it easy to build addon operators that follow the
+best practices we identify as part of this work.  For example, we expect that
+most addons will be declarative, and likely be specified as part of a "cluster
+bundle", so we will make it easy to build basic addon operators that follow
+these patterns.
+
+We hope that components will choose to maintain their own operators, encoding
+their knowledge of how best to operate their addon.
+
+
+## Motivation
+
+Addons are components that are managed alongside the lifecycle of the cluster.
+They are often tied to or dependent on the configuration of other cluster
+components.  Management of these components has proved complicated.  Our
+existing solution in the form of the bash addon-manager has many known
+shortcomings and is not widely adopted.  As we focus more development outside of
+the kubernetes/kubernetes repo, we expect more addon components of greater
+complexity.  This is one of the long-standing backlog items for
+sig-cluster-lifecycle.
+
+Use of operators is now generally accepted, and the benefits to other
+applications are generally recognized.  We aim to bring the benefits of
+operators to addons also.
+
+### Goals
+
+* Explore the use of operators for managing addons
+* Create patterns, libraries & tooling so that addons are of high quality,
+  consistent in their API surface (common fields on CRDs, use of Application
+  CRD, consistent labeling of created resources), yet are easy to build.
+* Build addons for the basic set of components, acting as a quality reference
```

Comment:
How does bundles fit under these goals?

Reply:
So ideally each addon operator references a set of manifests somewhere, rather than baking them into the image, so we don't _always_ have to update the operator (just when the underlying change is non-trivial).

Those manifests could be in any format really.  They could be simple yaml files.  I do think we should recommend _a_ format.

If we actually pick bundles as our format, this gives us a bit of metadata when pulling from https, but because bundles are CRDs, this also gives us the ability to fetch those manifests from the cluster rather than relying on some https server, i.e. it's "half" of airgap support.

Reply:
Doesn't that mean that the manifests themselves are the artifacts and the image should be embedded in them?

Reply:
Please not files on a disk that have to be rsync'ed.

Reply:
Not sure I follow @smarterclayton - manifests do specify an image name, but I'm missing where your going with that...

@bgrant0607 I don't think we would recommend files on a disk for any production configuration :-)